Email notification HibernateObjectRetrievalFailureException when associated task or project deleted

Email notification jobs throwing a HibernateObjectRetrievalFailureException when the notification's associated task or project has been deleted.

Schedule an email notification job (eg task assigned to resource)
Start scheduled job for notification
Allocate a resource to a task.
Delete task before the scheduled notification task runs
Review logs.

ERROR [09-January 11:15:05] [schedulerFactoryBean_Worker-9] org.quartz.core.JobRunShell  - Job EMAIL.TASK_ASSIGNED_TO_RESOURCE threw an unhandled Exception: 
org.springframework.orm.hibernate5.HibernateObjectRetrievalFailureException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]; nested exception is org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]
	at org.springframework.orm.hibernate5.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:197)
	at org.springframework.orm.hibernate5.HibernateExceptionTranslator.convertHibernateAccessException(HibernateExceptionTranslator.java:57)
	at org.springframework.orm.hibernate5.HibernateExceptionTranslator.translateExceptionIfPossible(HibernateExceptionTranslator.java:44)
	at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:59)
	at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:213)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:147)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy97.getAllByType(Unknown Source)
	at org.libreplan.web.email.EmailNotificationModel.getAllByType(EmailNotificationModel.java:68)
	at sun.reflect.GeneratedMethodAccessor528.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy211.getAllByType(Unknown Source)
	at org.libreplan.importers.notifications.realization.SendEmailOnTaskAssignedToResource.sendEmail(SendEmailOnTaskAssignedToResource.java:67)
	at sun.reflect.GeneratedMethodAccessor530.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy236.sendEmail(Unknown Source)
	at org.libreplan.importers.notifications.jobs.SendEmailOnTaskAssignedToResourceJob.executeInternal(SendEmailOnTaskAssignedToResourceJob.java:48)
	at org.springframework.scheduling.quartz.QuartzJobBean.execute(QuartzJobBean.java:75)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]
	at org.hibernate.boot.internal.StandardEntityNotFoundDelegate.handleEntityNotFound(StandardEntityNotFoundDelegate.java:28)
	at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:227)
	at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:278)
	at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:121)
	at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:89)
	at org.hibernate.internal.SessionImpl.fireLoad(SessionImpl.java:1142)
	at org.hibernate.internal.SessionImpl.internalLoad(SessionImpl.java:1025)
	at org.hibernate.type.EntityType.resolveIdentifier(EntityType.java:632)
	at org.hibernate.type.EntityType.resolve(EntityType.java:424)
	at org.hibernate.engine.internal.TwoPhaseLoad.doInitializeEntity(TwoPhaseLoad.java:151)
	at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntity(TwoPhaseLoad.java:125)
	at org.hibernate.loader.Loader.initializeEntitiesAndCollections(Loader.java:1139)
	at org.hibernate.loader.Loader.processResultSet(Loader.java:998)
	at org.hibernate.loader.Loader.doQuery(Loader.java:936)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:342)
	at org.hibernate.loader.Loader.doList(Loader.java:2622)
	at org.hibernate.loader.Loader.doList(Loader.java:2605)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2434)
	at org.hibernate.loader.Loader.list(Loader.java:2429)
	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:109)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1787)
	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:363)
	at org.libreplan.business.email.daos.EmailNotificationDAO.getAllByType(EmailNotificationDAO.java:50)
	at sun.reflect.GeneratedMethodAccessor529.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136)
	... 33 more
ERROR [09-January 11:15:05] [schedulerFactoryBean_Worker-9] org.quartz.core.ErrorLogger  - Job (EMAIL.TASK_ASSIGNED_TO_RESOURCE threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: org.springframework.orm.hibernate5.HibernateObjectRetrievalFailureException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]; nested exception is org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.springframework.orm.hibernate5.HibernateObjectRetrievalFailureException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]; nested exception is org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]
	at org.springframework.orm.hibernate5.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:197)
	at org.springframework.orm.hibernate5.HibernateExceptionTranslator.convertHibernateAccessException(HibernateExceptionTranslator.java:57)
	at org.springframework.orm.hibernate5.HibernateExceptionTranslator.translateExceptionIfPossible(HibernateExceptionTranslator.java:44)
	at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:59)
	at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:213)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:147)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy97.getAllByType(Unknown Source)
	at org.libreplan.web.email.EmailNotificationModel.getAllByType(EmailNotificationModel.java:68)
	at sun.reflect.GeneratedMethodAccessor528.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy211.getAllByType(Unknown Source)
	at org.libreplan.importers.notifications.realization.SendEmailOnTaskAssignedToResource.sendEmail(SendEmailOnTaskAssignedToResource.java:67)
	at sun.reflect.GeneratedMethodAccessor530.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy236.sendEmail(Unknown Source)
	at org.libreplan.importers.notifications.jobs.SendEmailOnTaskAssignedToResourceJob.executeInternal(SendEmailOnTaskAssignedToResourceJob.java:48)
	at org.springframework.scheduling.quartz.QuartzJobBean.execute(QuartzJobBean.java:75)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 more
Caused by: org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.libreplan.business.planner.entities.TaskElement#285]
	at org.hibernate.boot.internal.StandardEntityNotFoundDelegate.handleEntityNotFound(StandardEntityNotFoundDelegate.java:28)
	at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:227)
	at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:278)
	at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:121)
	at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:89)
	at org.hibernate.internal.SessionImpl.fireLoad(SessionImpl.java:1142)
	at org.hibernate.internal.SessionImpl.internalLoad(SessionImpl.java:1025)
	at org.hibernate.type.EntityType.resolveIdentifier(EntityType.java:632)
	at org.hibernate.type.EntityType.resolve(EntityType.java:424)
	at org.hibernate.engine.internal.TwoPhaseLoad.doInitializeEntity(TwoPhaseLoad.java:151)
	at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntity(TwoPhaseLoad.java:125)
	at org.hibernate.loader.Loader.initializeEntitiesAndCollections(Loader.java:1139)
	at org.hibernate.loader.Loader.processResultSet(Loader.java:998)
	at org.hibernate.loader.Loader.doQuery(Loader.java:936)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:342)
	at org.hibernate.loader.Loader.doList(Loader.java:2622)
	at org.hibernate.loader.Loader.doList(Loader.java:2605)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2434)
	at org.hibernate.loader.Loader.list(Loader.java:2429)
	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:109)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1787)
	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:363)
	at org.libreplan.business.email.daos.EmailNotificationDAO.getAllByType(EmailNotificationDAO.java:50)
	at sun.reflect.GeneratedMethodAccessor529.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136)
	... 33 more

--- a/libreplan-business/src/main/java/org/libreplan/business/email/daos/EmailNotificationDAO.java
+++ b/libreplan-business/src/main/java/org/libreplan/business/email/daos/EmailNotificationDAO.java
@@ -23,8 +23,12 @@ import org.hibernate.criterion.Restrictions;
 import org.libreplan.business.common.daos.GenericDAOHibernate;
 import org.libreplan.business.email.entities.EmailNotification;
 import org.libreplan.business.email.entities.EmailTemplateEnum;
+import org.libreplan.business.orders.entities.Order;
+import org.libreplan.business.planner.entities.TaskElement;
 import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
 
+//import java.util.ArrayList;
 import java.util.List;
 
 /**
@@ -47,6 +51,22 @@ public class EmailNotificationDAO
         return getSession()
                 .createCriteria(EmailNotification.class)
                 .add(Restrictions.eq("type", enumeration))
+                .list();
+    }
+
+    @Override
+    public List<EmailNotification> getAllByProject(TaskElement taskElement) {
+       return getSession()
+               .createCriteria(EmailNotification.class)
+               .add(Restrictions.eq("project", taskElement))
+               .list();
+    }
+
+    @Override
+    public List<EmailNotification> getAllByTask(TaskElement taskElement) {
+        return getSession()
+                .createCriteria(EmailNotification.class)
+                .add(Restrictions.eq("task", taskElement))
                 .list();
     }
 
@@ -87,6 +107,28 @@ public class EmailNotificationDAO
                 .createCriteria(EmailNotification.class)
                 .add(Restrictions.eq("id", notification.getId()))
                 .uniqueResult() == null;
+    }
+
+    @Override
+    public boolean deleteByProject(TaskElement taskElement) {
+        List<EmailNotification> notifications = getAllByProject(taskElement);
+
+        for (Object item : notifications){
+            getSession().delete(item);
+        }
+
+        return getAllByProject(taskElement).isEmpty();
+    }
+
+    @Override
+    public boolean deleteByTask(TaskElement taskElement) {
+        List<EmailNotification> notifications = getAllByTask(taskElement);
+
+        for (Object item : notifications){
+            getSession().delete(item);
+        }
+
+        return getAllByTask(taskElement).isEmpty();
     }
 
 }

--- a/libreplan-business/src/main/java/org/libreplan/business/email/daos/IEmailNotificationDAO.java
+++ b/libreplan-business/src/main/java/org/libreplan/business/email/daos/IEmailNotificationDAO.java
@@ -22,6 +22,8 @@ package org.libreplan.business.email.daos;
 import org.libreplan.business.common.daos.IGenericDAO;
 import org.libreplan.business.email.entities.EmailNotification;
 import org.libreplan.business.email.entities.EmailTemplateEnum;
+import org.libreplan.business.orders.entities.Order;
+import org.libreplan.business.planner.entities.TaskElement;
 
 import java.util.List;
 
@@ -36,9 +38,18 @@ public interface IEmailNotificationDAO extends IGenericDAO<EmailNotification, Lo
 
     List<EmailNotification> getAllByType(EmailTemplateEnum enumeration);
 
+    List<EmailNotification> getAllByProject(TaskElement taskElement);
+
+    List<EmailNotification> getAllByTask(TaskElement taskElement);
+
     boolean deleteAll();
 
     boolean deleteAllByType(EmailTemplateEnum enumeration);
 
     boolean deleteById(EmailNotification notification);
+
+    boolean deleteByProject(TaskElement taskElement);
+
+    boolean deleteByTask(TaskElement taskElement);
+
 }

--- a/libreplan-business/src/main/java/org/libreplan/business/orders/daos/OrderElementDAO.java
+++ b/libreplan-business/src/main/java/org/libreplan/business/orders/daos/OrderElementDAO.java
@@ -39,6 +39,7 @@ import org.hibernate.Query;
 import org.hibernate.criterion.Restrictions;
 import org.libreplan.business.common.daos.IntegrationEntityDAO;
 import org.libreplan.business.common.exceptions.InstanceNotFoundException;
+import org.libreplan.business.email.daos.IEmailNotificationDAO;
 import org.libreplan.business.expensesheet.daos.IExpenseSheetLineDAO;
 import org.libreplan.business.labels.entities.Label;
 import org.libreplan.business.orders.entities.Order;
@@ -46,6 +47,7 @@ import org.libreplan.business.orders.entities.OrderElement;
 import org.libreplan.business.orders.entities.SchedulingDataForVersion;
 import org.libreplan.business.orders.entities.TaskSource;
 import org.libreplan.business.planner.daos.ITaskSourceDAO;
+import org.libreplan.business.planner.entities.TaskElement;
 import org.libreplan.business.resources.entities.Criterion;
 import org.libreplan.business.templates.entities.OrderElementTemplate;
 import org.libreplan.business.workingday.EffortDuration;
@@ -83,6 +85,9 @@ public class OrderElementDAO extends IntegrationEntityDAO<OrderElement> implemen
 
     @Autowired
     private ITaskSourceDAO taskSourceDAO;
+
+    @Autowired
+    private IEmailNotificationDAO emailNotificationDAO;
 
     @Override
     public List<OrderElement> findWithoutParent() {
@@ -142,6 +147,9 @@ public class OrderElementDAO extends IntegrationEntityDAO<OrderElement> implemen
     @Override
     public void remove(Long id) throws InstanceNotFoundException {
         OrderElement orderElement = find(id);
+
+        removeNotifications(orderElement);
+
         removeTaskSourcesFor(this.taskSourceDAO, orderElement);
 
         for (WorkReport each : getWorkReportsPointingTo(orderElement)) {
@@ -149,6 +157,17 @@ public class OrderElementDAO extends IntegrationEntityDAO<OrderElement> implemen
         }
 
         super.remove(id);
+    }
+
+    public void removeNotifications(OrderElement orderElement) {
+
+        List<SchedulingDataForVersion> allVersions = orderElement.getSchedulingDataForVersionFromBottomToTop();
+        for (TaskSource each : taskSourcesFrom(allVersions)) {
+            TaskElement taskElement = each.getTask();
+            if ( taskElement != null) {
+                emailNotificationDAO.deleteByTask(taskElement);
+            }
+        }
     }
 
     public static void removeTaskSourcesFor(ITaskSourceDAO taskSourceDAO, OrderElement orderElement)

--- a/libreplan-webapp/src/main/java/org/libreplan/web/email/EmailNotificationModel.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/web/email/EmailNotificationModel.java
@@ -22,6 +22,7 @@ package org.libreplan.web.email;
 import org.libreplan.business.common.exceptions.ValidationException;
 import org.libreplan.business.email.daos.IEmailNotificationDAO;
 import org.libreplan.business.email.entities.EmailTemplateEnum;
+import org.libreplan.business.orders.entities.Order;
 import org.libreplan.business.email.entities.EmailNotification;
 
 import org.libreplan.business.planner.entities.TaskElement;
@@ -70,6 +71,17 @@ public class EmailNotificationModel implements IEmailNotificationModel {
 
     @Override
     @Transactional
+    public List<EmailNotification> getAllByProject(TaskElement taskElement) {
+        return emailNotificationDAO.getAllByProject(taskElement);
+    }
+    @Override
+    @Transactional
+    public List<EmailNotification> getAllByTask(TaskElement taskElement) {
+        return emailNotificationDAO.getAllByTask(taskElement);
+    }
+
+    @Override
+    @Transactional
     public boolean deleteAll() {
         return emailNotificationDAO.deleteAll();
     }
@@ -83,6 +95,18 @@ public class EmailNotificationModel implements IEmailNotificationModel {
     @Transactional
     public boolean deleteById(EmailNotification notification){
         return emailNotificationDAO.deleteById(notification);
+    }
+
+    @Override
+    @Transactional
+    public boolean deleteByProject(TaskElement taskElement) {
+        return emailNotificationDAO.deleteByProject(taskElement);
+    }
+
+    @Override
+    @Transactional
+    public boolean deleteByTask(TaskElement taskElement) {
+        return emailNotificationDAO.deleteByTask(taskElement);
     }
 
     @Override
@@ -110,11 +134,12 @@ public class EmailNotificationModel implements IEmailNotificationModel {
         this.emailNotification.setProject(project);
     }
 
-
+    @Override
     public EmailNotification getEmailNotification() {
         return emailNotification;
     }
 
+    @Override
     public void setNewObject(){
         this.emailNotification = new EmailNotification();
     }

--- a/libreplan-webapp/src/main/java/org/libreplan/web/email/IEmailNotificationModel.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/web/email/IEmailNotificationModel.java
@@ -21,6 +21,7 @@ package org.libreplan.web.email;
 
 import org.libreplan.business.common.exceptions.ValidationException;
 import org.libreplan.business.email.entities.EmailTemplateEnum;
+import org.libreplan.business.orders.entities.Order;
 import org.libreplan.business.email.entities.EmailNotification;
 import org.libreplan.business.planner.entities.TaskElement;
 import org.libreplan.business.resources.entities.Resource;
@@ -41,11 +42,19 @@ public interface IEmailNotificationModel {
 
     List<EmailNotification> getAllByType(EmailTemplateEnum enumeration);
 
+    List<EmailNotification> getAllByProject(TaskElement taskElement);
+
+    List<EmailNotification> getAllByTask(TaskElement taskElement);
+
     boolean deleteAll();
 
     boolean deleteAllByType(EmailTemplateEnum enumeration);
 
     boolean deleteById(EmailNotification notification);
+
+    boolean deleteByProject(TaskElement taskElement);
+
+    boolean deleteByTask(TaskElement taskElement);
 
     void setType(EmailTemplateEnum type);
 
@@ -60,4 +69,5 @@ public interface IEmailNotificationModel {
     EmailNotification getEmailNotification();
 
     void setNewObject();
+
 }

--- a/libreplan-webapp/src/main/java/org/libreplan/web/orders/OrderModel.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/web/orders/OrderModel.java
@@ -47,6 +47,7 @@ import org.libreplan.business.common.daos.IConfigurationDAO;
 import org.libreplan.business.common.entities.Configuration;
 import org.libreplan.business.common.entities.EntityNameEnum;
 import org.libreplan.business.common.exceptions.InstanceNotFoundException;
+import org.libreplan.business.email.entities.EmailNotification;
 import org.libreplan.business.externalcompanies.daos.IExternalCompanyDAO;
 import org.libreplan.business.externalcompanies.entities.EndDateCommunication;
 import org.libreplan.business.externalcompanies.entities.ExternalCompany;
@@ -60,6 +61,7 @@ import org.libreplan.business.orders.entities.OrderElement;
 import org.libreplan.business.orders.entities.OrderLineGroup;
 import org.libreplan.business.orders.entities.OrderStatusEnum;
 import org.libreplan.business.planner.entities.PositionConstraintType;
+import org.libreplan.business.planner.entities.TaskElement;
 import org.libreplan.business.qualityforms.daos.IQualityFormDAO;
 import org.libreplan.business.qualityforms.entities.QualityForm;
 import org.libreplan.business.requirements.entities.DirectCriterionRequirement;
@@ -84,6 +86,7 @@ import org.libreplan.business.users.entities.UserRole;
 import org.libreplan.web.calendars.BaseCalendarModel;
 import org.libreplan.web.common.IntegrationEntityModel;
 import org.libreplan.web.common.concurrentdetection.OnConcurrentModification;
+import org.libreplan.web.email.IEmailNotificationModel;
 import org.libreplan.web.logs.IIssueLogModel;
 import org.libreplan.web.logs.IRiskLogModel;
 import org.libreplan.web.orders.files.IOrderFileModel;
@@ -180,6 +183,9 @@ public class OrderModel extends IntegrationEntityModel implements IOrderModel {
 
     @Autowired
     private IIssueLogModel issueLogModel;
+
+    @Autowired
+    private IEmailNotificationModel emailNotificationModel;
 
     private List<Order> orderList = new ArrayList<>();
 
@@ -522,6 +528,8 @@ public class OrderModel extends IntegrationEntityModel implements IOrderModel {
     public void remove(Order detachedOrder) {
         Order order = orderDAO.findExistingEntity(detachedOrder.getId());
 
+        removeNotifications(order);
+
         removeFiles(order);
 
         removeLogs(order);
@@ -540,6 +548,18 @@ public class OrderModel extends IntegrationEntityModel implements IOrderModel {
 
     private void removeFiles(Order order) {
         orderFileModel.findByParent(order).forEach(orderFile -> orderFileModel.delete(orderFile));
+    }
+
+    private void removeNotifications(Order order) {
+
+        for ( Scenario scenario: currentAndDerivedScenarios()) {
+            order.useSchedulingDataFor(scenario.getOrderVersion(order));
+
+            TaskElement taskElement = order.getTaskElement();
+            if ( taskElement != null) {
+                emailNotificationModel.deleteByProject(taskElement);
+            }
+        }
     }
 
     private void removeVersions(Order order) {

--- a/libreplan-webapp/src/test/java/org/libreplan/web/test/ws/email/EmailTest.java
+++ b/libreplan-webapp/src/test/java/org/libreplan/web/test/ws/email/EmailTest.java
@@ -2,7 +2,9 @@ package org.libreplan.web.test.ws.email;
 
 
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import org.libreplan.business.common.Registry;
@@ -165,6 +167,20 @@ public class EmailTest {
         emailNotificationDAO.deleteAll();
 
         assertTrue(EmailConnectionValidator.exceptionType instanceof MessagingException);
+    }
+
+    @Test
+    @Transactional
+    public void testDDeleteEmailNotification() {
+        EmailTemplate emailTemplate = createEmailTemplate();
+        emailTemplateDAO.save(emailTemplate);
+
+        EmailNotification emailNotification = createEmailNotification();
+        emailNotificationDAO.save(emailNotification);
+
+        emailTemplateDAO.delete(emailTemplate);
+        boolean result = emailNotificationDAO.deleteByProject(emailNotification.getProject());
+        assertTrue(result);
     }
 
     private EmailTemplate createEmailTemplate() {

